x/concentrated-liquidity/pool.go

```diff
@@ -159,28 +142,32 @@ func (k Keeper) CalcOutAmtGivenIn(ctx sdk.Context,
 
 	// TODO: This should be GT 0 but some instances have very small remainder
 	// need to look into fixing this
-	for swapState.amountSpecifiedRemaining.GT(sdk.NewDecWithPrec(1, 6)) {
+	for swapState.amountSpecifiedRemaining.GT(sdk.ZeroDec()) && !swapState.sqrtPrice.Equal(sqrtPriceLimit) {
 		nextTick, ok := k.NextInitializedTick(ctx, poolId, swapState.tick.Int64(), zeroForOne)
 		if !ok {
-			return sdk.Coin{}, sdk.ZeroDec(), sdk.ZeroInt(), sdk.ZeroDec(), fmt.Errorf("there are no more ticks initialized to fill the swap")
+			return sdk.Coin{}, sdk.Coin{}, sdk.Int{}, sdk.Dec{}, fmt.Errorf("there are no more ticks initialized to fill the swap")
 		}
 
 		nextSqrtPrice, err := k.tickToSqrtPrice(sdk.NewInt(nextTick))
 		if err != nil {
-			return sdk.Coin{}, sdk.ZeroDec(), sdk.ZeroInt(), sdk.ZeroDec(), fmt.Errorf("could not convert next tick (%v) to nextSqrtPrice", sdk.NewInt(nextTick))
+			return sdk.Coin{}, sdk.Coin{}, sdk.Int{}, sdk.Dec{}, fmt.Errorf("could not convert next tick (%v) to nextSqrtPrice", sdk.NewInt(nextTick))
+		}
+
+		var sqrtPriceTarget sdk.Dec
+		if zeroForOne && nextSqrtPrice.LT(sqrtPriceLimit) {
+			sqrtPriceTarget = sqrtPriceLimit
+		} else {
+			sqrtPriceTarget = nextSqrtPrice
```

Comment:
I don't think this is entirely right because `if !zeroForOne && nextSqrtPrice.GT(sqrtPriceLimit)` then `sqrtPriceTarget = sqrtPriceLimit`

Reply:
good catch! added fix